compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/ir/IrUtils.kt

```diff
@@ -417,3 +419,25 @@ inline fun IrElement.hasChild(crossinline block: (IrElement) -> Boolean): Boolea
     }, null)
     return result
 }
+
+// Map declarations to original declarations before lowering.
+private val IrDeclaration.original: IrDeclaration
+    get() = (this as? IrAttributeContainer)?.attributeOwnerId as? IrDeclaration ?: this
+
+// Declarations in the scope of an externally visible inline function are implicitly part of the
+// public ABI of a Kotlin module. This function returns the visibility of a containing inline function
+// (determined *before* lowering), or null if the given declaration is not in the scope of an inline function.
+val IrDeclaration.inlineScopeVisibility: DescriptorVisibility?
```

Comment:
Could you also add (abi?) test for next case:
```
var result = "fail"

inline fun foo(crossinline s: () -> String) {

    object {
        private inline fun test(crossinline z: () -> String) {
            {
                result = z()
            }()
        }

        fun foo() {
            test { s() }
        }
    }.foo()
}

fun box() {
    foo {
        "OK"
    }
}

fun main() {
    box()
    println(result)
}
```

Reply:
and without inner object regeneration (I've also updated snippet above a bit):
```
var result = "fail"

inline fun foo(crossinline s: () -> String) {

    object {
        private inline fun test(crossinline z: () -> String) {
            result = object {
                fun run(): String {
                    return "O"
                }
            }.run() + z()
        }

        fun foo() {
            test { s() }
        }
    }.foo()
}

fun box() {
    foo {
        "K"
    }
}

fun main() {
    box()
    println(result)
}
```

Reply:
I'm consistently amazed by how quickly you can come up with these convoluted examples.

This was not handled correctly, as I mistakenly assumed that anonymous objects inside of private inline functions could not escape. Your second example shows that this is not the case. I've fixed this problem and added the tests (in a new commit to make it easier to review).

On a side note, the first test is also broken in the legacy jvm-abi-gen implementation.

Reply:
> I'm consistently amazed by how quickly you can come up with these convoluted examples.

Thank you, nice to hear!
But still some cases could be missed anyway, and I've forgot about next ones
```
class A {
    private class B {
        inline fun test(crossinline s: () -> Unit) {
            val z = 1;
            object {
                fun run() {
                    s()
                }
            }.run()
        }
    }
}

private class C {
    inline fun test(crossinline s: () -> Unit) {
        val z = 1;
        object {
            fun run() {
                s()
            }
        }.run()
    }
}
```